### PR TITLE
feat: Reentrancy Guard

### PR DESCRIPTION
**Pull request by ermvrs**
Fix: https://github.com/keep-starknet-strange/satoru/issues/244

# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build-related changes
- Documentation content changes
- Testing
- Other (please describe):

## What is the current behavior?

Resolves: #NA

## What is the new behavior?

-
-
-

### CODE REVIEW

File: src/utils/global_reentrancy_guard.cairo [hunk 1]
```diff
@@ -5,12 +5,21 @@
...
 
 /// Modifier to avoid reentrancy.
-fn non_reentrant_before(data_store: IDataStoreDispatcher) { // TODO
+fn non_reentrant_before(data_store: IDataStoreDispatcher) { 
+    let status = match data_store.get_bool('REENTRANCY_GUARD_STATUS') {
+        Option::Some(v) => v,
+        Option::None(()) => false,
+    }; // Read key from Data Store
+
+    assert(!status, 'ReentrancyGuard: reentrant call');
```
Comment (abdelhamidbakhta) [inline src/utils/global_reentrancy_guard.cairo:17]: nit: error message in a constant in an error file, following conventions of other error messages

File: src/utils/global_reentrancy_guard.cairo [hunk 2]
```diff
@@ -5,12 +5,21 @@
...
 // Satoru imports
-use satoru::data::data_store::{IDataStoreDispatcher};
+use satoru::data::data_store::{IDataStoreDispatcher, IDataStoreDispatcherTrait};
 
 /// Modifier to avoid reentrancy.
-fn non_reentrant_before(data_store: IDataStoreDispatcher) { // TODO
+fn non_reentrant_before(data_store: IDataStoreDispatcher) { 
+    let status = match data_store.get_bool('REENTRANCY_GUARD_STATUS') {
+        Option::Some(v) => v,
+        Option::None(()) => false,
```
Comment (abdelhamidbakhta) [inline src/utils/global_reentrancy_guard.cairo:14]: double `(` `)` can be removed

File: src/utils/global_reentrancy_guard.cairo [hunk 3]
```diff
@@ -5,12 +5,21 @@
...
+    }; // Read key from Data Store
+
+    assert(!status, 'ReentrancyGuard: reentrant call');
+
+    data_store.set_bool('REENTRANCY_GUARD_STATUS', true);
```
Comment (abdelhamidbakhta) [inline src/utils/global_reentrancy_guard.cairo:19]: use a const for `REENTRANCY_GUARD_STATUS` value

File: tests/utils/test_reentrancy_guard.cairo [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        Option::Some(v) => v,
+        Option::None(()) => false,
+    };
+
+    assert(!initial_value, 'Initial value wrong'); // Initial value should be false.
+
+    non_reentrant_before(data_store); // Sets value to true
+    
+    // Gets value after non_reentrant_before call
+    let entraant: bool = match data_store.get_bool('REENTRANCY_GUARD_STATUS') {
```
Comment (abdelhamidbakhta) [inline tests/utils/test_reentrancy_guard.cairo:80]: typo: `entrant`

**Review by abdelhamidbakhta — CHANGES_REQUESTED** [review]
left some comments

File: tests/utils/test_reentrancy_guard.cairo (hunk 4 above)
Comment (ermvrs) [reply tests/utils/test_reentrancy_guard.cairo]: https://github.com/keep-starknet-strange/satoru/pull/258/commits/8c0e057f3d39e1238129bd722123b34ac5d961c9

File: src/utils/global_reentrancy_guard.cairo [hunk 5]
```diff
@@ -5,12 +5,24 @@
 use starknet::ContractAddress;
 
 // Satoru imports
-use satoru::data::data_store::{IDataStoreDispatcher};
+use satoru::data::data_store::{IDataStoreDispatcher, IDataStoreDispatcherTrait};
+use satoru::utils::error::{ReentrancyGuardError};
```
Comment (zarboq) [inline src/utils/global_reentrancy_guard.cairo:9]:
```suggestion
use satoru::utils::error::ReentrancyGuardError;
```

**Pull request merged**